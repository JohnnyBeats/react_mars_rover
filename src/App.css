:root {
  --50-dark: rgba(0, 0, 0, .5);
  --dark-font-color: #ffffff;
}

html,
.App,
.cover,
.rover,
body {
  height: 100%;
  width: 100%;
}

body{
  background-image: url(Images/floor.jpg);
  position: absolute;
}

input[type='button'], .button{
  background: #121212db;
  color: var(--dark-font-color);
  border: none;
  padding: 1em;
  box-shadow: 0 0 0 1px var(--50-dark);
}

textarea[name='input']{
  width: 50vw;
  height: 50vh;
  background: rgba(0,0,0,0.8);
  color: var(--dark-font-color);
  border-color: #000000;
  resize: none;
  padding: 2em;
}

button.openForm{
  position: fixed;
  right: 20px;
  bottom: 20px;
  z-index: 2;
}

.App {
  text-align: center;
}

.messageContainer{
  position: fixed;
  right: 0;
  top: 0;
  width: 330px;
  height: calc(100% - 90px);
  overflow-x: hidden;
  overflow-y: auto;
}

.messageContainer .message{
  width: 300px;
  background: #121212db;
  box-shadow: 0 0 3px 1px rgba(0,0,0,.6);
  border: 1px solid #565656;
  color: #fff;
  right: 0;
  position: relative;
  margin: 1em;
}

.messageContainer .message .text{
  max-width: calc(100% - 3em);
  padding: 1em;
  text-align: left;
}

.messageContainer .message .close{
  position: relative;
  float: right;
  cursor: pointer;
  padding: 1em;
}

.cover{
  overflow: hidden;
  position: absolute;
  z-index: 100;
  background: #121212 url(Images/inputCoverBackground.jpg);
  background-size: cover;
}

.cover h1{
  opacity: .6;
  font-size: 4vw;
}

.cover form{
  display: block;
}

.cover.hidden{
  height: 0;
  width: 0;
  display: block;
}

.grid{
  display: table;
  border-style: solid;
  border-color: var(--50-dark);
  border-width: 1px 0 0 1px;
  margin: 16px auto 16px auto;
  color: #000;
  font-weight: bolder;
}

.grid.hidden{
  display: none;
}

.column{
  display: table-cell;
  border-style: solid;
  border-color: var(--50-dark);
  border-width: 0 1px 1px 0;
  width: 5vw;
  height: 5vw;
  min-width: 5vw;
  min-height: 5vw;
  overflow: hidden;
  text-overflow: ellipsis;
  color: var(--50-dark);
  vertical-align: middle;
  font-size: 2em;
  font-weight: 400;
}

.columnCoordinate{
  transform: rotate(45deg);
  display: block;
  color: rgba(0,0,0,.7);
}

.rover{
  position: relative;
  background: url(Images/rover.png);
  background-size: cover;
  box-sizing: border-box;
  animation: ease;
  -webkit-transition-duration: 1s;
  -webkit-transition-property: -webkit-transform;
}

.rover.active{
  background-color: #ffff003d;
}

@keyframes welcome {
  from {
    font-size: 14vw;
    opacity: 0;
  }
  25%  {
    opacity: 1;
  }
  50%  {
    font-size: 14vw;
  }
  to {
    font-size: 4vw;
    opacity: .6;
  }
}

@keyframes slideUp {
  from {
    height: 100%;
    width: 100%;
  }
  to {
    height: 0;
    width: 0;
  }
}

@keyframes appear {
  from {
    opacity: 0;
  }
  80%  {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

@media (prefers-reduced-motion: no-preference) {
  .cover h1 {
    animation: welcome 3s linear;
  }
  .cover form{
    animation: appear 4s linear;
  }
  .cover.hidden{
    animation: slideUp .3s linear;
  }
}

@media screen and (max-width: 1600px) {
  .column {
    font-size: 1.5em;
    width: 10vw;
    height: 10vw;
    min-width: 10vw;
    min-height: 10vw;
  }
}

@media screen and (max-width: 1200px) {
  .column {
    width: 15vw;
    height: 15vw;
    min-width: 15vw;
    min-height: 15vw;
  }
}

@media screen and (max-width: 768px) {
  .columnCoordinate{
    color: rgba(0,0,0,1);
  }

  .cover h1 {
    font-size: 8vw;
    color: rgba(255, 255, 255, 0.3);
  }

  textarea[name='input'] {
    width: 75vw;
    height: 60vh;
  }
}

@media screen and (max-width: 768px) and (orientation: landscape) {
  .cover h1 {
    font-size: 4vw;
  }

  textarea[name='input'] {
    height: 50vh;
  }
}