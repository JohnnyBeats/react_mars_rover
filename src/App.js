import React from 'react';
import './App.css';
import Grid from './Components/Grid.js'
import SnackMessage from "./Components/SnackMessage";
import InputForm from "./Components/InputForm";
import GameModeSelection from "./Components/GameModeSelection";
import {RoverEngineMultiplayer, RoverTeam} from "./Classes/Rover";
import firebase from "./firebase.js";
import GameModeEnum from "./Classes/GameModeEnum";
import DirectionEnum from "./Classes/DirectionEnum";

class App extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            showCover: true,
            gameMode: undefined,
            roverTeam: new RoverTeam(),
            gridSize: [0,0],
            roverIndex: 0,
            roverActiveInterval: undefined,
            messages: []
        };

        this.setAppData = this.setAppData.bind(this);
        this.setGameMode = this.setGameMode.bind(this);
        this.addMessage = this.addMessage.bind(this);
        this.resetRovers = this.resetRovers.bind(this);
        this.startRovers = this.startRovers.bind(this);
        this.removeMessage = this.removeMessage.bind(this);
    }

    addMessage = (message) => {
        this.state.messages.push(message);
        this.setState({
            messages: this.state.messages
        });
    };

    removeMessage = (event) => {
        this.state.messages.splice(event, 1);
        this.setState({
            messages:  this.state.messages
        });
    };

    setAppData = (showCover, roverTeam, gridSize) => {
        this.setState({
            showCover: showCover,
            roverTeam: roverTeam,
            gridSize: gridSize
        });

        this.startRovers();
    };

    setGameMode = (gameMode) => {
        this.setState({
            gameMode: gameMode
        }, () =>{
            if(this.state.gameMode === GameModeEnum.MULTIPLAYER){
                const playersRef = firebase.database().ref("Game/Players");

                const randomGridNumber = (max) => {
                    return Math.floor(Math.random() * (max + 1));
                }

                playersRef.on("value", (snapshot) => {
                    let players = snapshot.val();
                    let roverTeam = new RoverTeam();
                    let maxGridInput = [8, 4];

                    players.map((player, index) => {
                        let position = [randomGridNumber(maxGridInput[0]), randomGridNumber(maxGridInput[1])];
                        let direction = DirectionEnum.NORTH;
                        let commands = "";
                        let playerRoverEngine = new RoverEngineMultiplayer(position, direction, commands, maxGridInput, roverTeam, index, player.name);

                        roverTeam.rovers.push(playerRoverEngine);

                        const playerCommandRef = firebase.database().ref("Game/Commands/" + index + "/command");

                        playerCommandRef.on("value", (snapshot) => {
                            let commandData = snapshot.val();

                            playerRoverEngine.processCommand(commandData.command);

                            this.setState({
                                roverTeam: this.state.roverTeam
                            });
                        });
                    });

                    this.setAppData(false, roverTeam, maxGridInput);
                });
            }
        });
    };

    resetRovers = () => {
        if(this.state.roverActiveInterval !== undefined){
            clearInterval(this.state.roverActiveInterval);
        }

        const roverTeam = new RoverTeam();
        this.setState({
            roverActiveInterval: undefined,
            showCover: true,
            roverTeam: roverTeam,
            gridSize: [0,0],
            roverIndex: 0
        });
    };

    startRovers = () => {
        //Interval here to make rover move visually for each command
        const roverActiveInterval = setInterval(() => {
            this.state.roverTeam.rovers[this.state.roverIndex].runNextCommand();

            //Update rovers here so UI can update
            this.setState({
                roverTeam: this.state.roverTeam
            });

            if(this.state.roverTeam.rovers[this.state.roverIndex].commandsComplete){
                this.addMessage("Rover " + (this.state.roverIndex + 1) + " has stopped on: " + this.state.roverTeam.rovers[this.state.roverIndex].getPosition());
                if(this.state.roverIndex < this.state.roverTeam.rovers.length - 1){
                    this.setState({
                        roverIndex: this.state.roverIndex + 1
                    });
                }else{
                    clearInterval(roverActiveInterval);
                    this.setState({
                        roverActiveInterval: undefined
                    });

                }
            }
        }, 2000);

        this.setState({
            roverActiveInterval: roverActiveInterval
        })
    };

    render() {
        return (
            <div className="App">
                <div data-testid="inputCover" className={this.state.showCover ? "cover": "cover hidden"}>
                    <h1> MARS ROVER </h1>
                    {this.state.gameMode === undefined &&
                        <GameModeSelection setGameMode={this.setGameMode} />
                    }
                    {this.state.gameMode === GameModeEnum.MULTIPLAYER &&
                        <h2>Waiting for players to connect...</h2>
                    }
                    {this.state.gameMode === GameModeEnum.SINGLE_PLAYER &&
                        <InputForm setAppData={this.setAppData}/>
                    }
                </div>
            {!this.state.showCover &&
                <button className="openForm button" onClick={this.resetRovers}>
                    Edit Commands
                </button>
            }
                <Grid showCover={this.state.showCover} roverTeam={this.state.roverTeam} gridSize={this.state.gridSize}/>
                <div className="messageContainer">
                    {this.state.messages.map((value, index) => {
                        return <SnackMessage key={"message-" + index} index={index} text={value} onClose={this.removeMessage}/>
                    })}
                </div>
            </div>
        )
    }
}

export default App;
