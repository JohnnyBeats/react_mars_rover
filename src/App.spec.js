import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import App from './App';

describe("App", () => {
  window.alert = jest.fn();

  it('renders according to spec with cover and form visible', () => {
    const {container} = render(<App/>);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe("Input Form", () => {

    it('Renders with form visible', () => {
      const {getByTestId} = render(<App/>);
      expect(getByTestId("inputForm")).toBeVisible();
    });

    it('loads rover grid with correct input', async () => {
      const {container, getByTestId, getAllByTestId, findByText} = render(<App/>);
      const startButton = getByTestId('startButton');
      const commandsInput = getByTestId('commandsInput');

      fireEvent.change(commandsInput, {
        target: {
          value: ('5 5\n1 2 N\nLMLMLMLMM\n3 3 E\nMMRMMRMRRM\n')
        }
      });

      // Click button
      fireEvent.click(startButton);
      window.alert.mockClear();

      // Wait for page to update with query text
      const maxCell = await findByText(/5,5/);
      expect(maxCell).toBeTruthy();

      //Check that we have two rovers
      const rovers = await getAllByTestId("rover");
      expect(rovers).toHaveLength(2);

      expect(container.firstChild).toMatchSnapshot();
    })

  });

});
