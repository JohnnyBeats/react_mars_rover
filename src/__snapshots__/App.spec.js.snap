// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Input Form loads rover grid with correct input 1`] = `
<div
  class="App"
>
  <div
    class="cover hidden"
    data-testid="inputCover"
  >
    <h1>
       MARS ROVER 
    </h1>
    <form
      data-testid="inputForm"
    >
      <textarea
        data-testid="commandsInput"
        name="input"
        placeholder="5 5
1 2 N
LMLMLMLMM
3 3 E
MMRMMRMRRM
"
      >
        5 5
1 2 N
LMLMLMLMM
3 3 E
MMRMMRMRRM

      </textarea>
      <br />
      <input
        data-testid="startButton"
        type="button"
        value="Start Rovers"
      />
    </form>
  </div>
  <button
    class="openForm button"
  >
    Edit Commands
  </button>
  <div
    class="grid"
  >
    <div
      class="row"
    >
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="0,5"
        >
          0,5
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="1,5"
        >
          1,5
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="2,5"
        >
          2,5
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="3,5"
        >
          3,5
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="4,5"
        >
          4,5
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="5,5"
        >
          5,5
        </span>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="0,4"
        >
          0,4
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="1,4"
        >
          1,4
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="2,4"
        >
          2,4
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="3,4"
        >
          3,4
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="4,4"
        >
          4,4
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="5,4"
        >
          5,4
        </span>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="0,3"
        >
          0,3
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="1,3"
        >
          1,3
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="2,3"
        >
          2,3
        </span>
      </div>
      <div
        class="column"
      >
        <div
          class="rover east inActive"
          data-testid="rover"
          style="transform: rotate(90deg);"
        />
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="4,3"
        >
          4,3
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="5,3"
        >
          5,3
        </span>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="0,2"
        >
          0,2
        </span>
      </div>
      <div
        class="column"
      >
        <div
          class="rover north inActive"
          data-testid="rover"
          style="transform: rotate(0deg);"
        />
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="2,2"
        >
          2,2
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="3,2"
        >
          3,2
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="4,2"
        >
          4,2
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="5,2"
        >
          5,2
        </span>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="0,1"
        >
          0,1
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="1,1"
        >
          1,1
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="2,1"
        >
          2,1
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="3,1"
        >
          3,1
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="4,1"
        >
          4,1
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="5,1"
        >
          5,1
        </span>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="0,0"
        >
          0,0
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="1,0"
        >
          1,0
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="2,0"
        >
          2,0
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="3,0"
        >
          3,0
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="4,0"
        >
          4,0
        </span>
      </div>
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="5,0"
        >
          5,0
        </span>
      </div>
    </div>
  </div>
  <div
    class="messageContainer"
  />
</div>
`;

exports[`App renders according to spec with cover and form visible single payer 1`] = `
<div
  class="App"
>
  <div
    class="cover"
    data-testid="inputCover"
  >
    <h1>
       MARS ROVER 
    </h1>
    <form
      data-testid="inputForm"
    >
      <textarea
        data-testid="commandsInput"
        name="input"
        placeholder="5 5
1 2 N
LMLMLMLMM
3 3 E
MMRMMRMRRM
"
      />
      <br />
      <input
        data-testid="startButton"
        type="button"
        value="Start Rovers"
      />
    </form>
  </div>
  <div
    class="grid hidden"
  >
    <div
      class="row"
    >
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="0,0"
        >
          0,0
        </span>
      </div>
    </div>
  </div>
  <div
    class="messageContainer"
  />
</div>
`;

exports[`App renders according to spec with game mode choices available 1`] = `
<div
  class="App"
>
  <div
    class="cover"
    data-testid="inputCover"
  >
    <h1>
       MARS ROVER 
    </h1>
    <div
      class="gameModeSelection"
    >
      <button
        class="gameMode button"
        data-testid="Multiplayer"
        value="Multiplayer"
      >
        Multiplayer
      </button>
      <br />
      <br />
      <button
        class="gameMode button"
        data-testid="SinglePlayer"
        value="SinglePlayer"
      >
        Single Player
      </button>
    </div>
  </div>
  <div
    class="grid hidden"
  >
    <div
      class="row"
    >
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="0,0"
        >
          0,0
        </span>
      </div>
    </div>
  </div>
  <div
    class="messageContainer"
  />
</div>
`;

exports[`App renders according to spec with waiting for players multiplayer 1`] = `
<div
  class="App"
>
  <div
    class="cover"
    data-testid="inputCover"
  >
    <h1>
       MARS ROVER 
    </h1>
    <h2>
      Waiting for players to connect...
    </h2>
  </div>
  <div
    class="grid hidden"
  >
    <div
      class="row"
    >
      <div
        class="column"
      >
        <span
          class="columnCoordinate"
          title="0,0"
        >
          0,0
        </span>
      </div>
    </div>
  </div>
  <div
    class="messageContainer"
  />
</div>
`;
